G-Earth/src/main/java/gearth/protocol/crypto/RC4.java

```diff
@@ -170,37 +143,56 @@ public byte[] rc4(byte[] buf) {
             state[x] = state[y];
             state[y] = tmp;
 
-            xorIndex = ((state[x] &0xff) + (state[y] & 0xff)) & 0xff;
-            result[i] = (byte)(buf[i] ^ state[xorIndex]);
+            xorIndex = ((state[x] & 0xff) + (state[y] & 0xff)) & 0xff;
+            result[i] = (byte) (data[offset + i] ^ state[xorIndex]);
         }
 
-        //this.x = lx;
-        //this.y = ly;
-
         return result;
     }
 
+    @Override
+    public byte[] decipher(byte[] data) {
+        return cipher(data);
+    }
+
+    @Override
+    public byte[] decipher(byte[] data, int offset, int length) {
+        return cipher(data, offset, length);
+    }
+
+    @Override
+    public byte[] getState () {
+        return state;
+    }
+
+    @Override
+    public int getQ() {
+        return x;
+    }
+
+    @Override
+    public int getJ() {
+        return y;
+    }
+
+    public RC4 deepCopy() {
+        return new RC4(Arrays.copyOf(state, 256), x, y);
+    }
+
     public boolean couldBeFresh() {
         return (x == 0 && y == 0);
     }
 
     public void undoRc4(byte[] buf) {
-
         byte tmp;
 
-        for (int i = buf.length - 1; i >= 0; i--) {
-
+        for (int i = 0; i < buf.length; i++) {
```

Comment:
I have no idea if this is right or wrong, but my linux g-memz thing worked as soon as i changed this back to the original line